### DISCUSSION TITLE
Settle on Github Workflow Conventions for Contributors

### DISCUSSION THREAD

**Discussion by njsch** [eduvis/eduvis.github.io#14] [General]
Hey @Coedice, as you must have already noticed, I've been a really bad dev and been committing everything to Main. It hasn't really mattered yet as I've been (virtually) the only contributor. How do you propose we do this?

I'm happy to create feature (enhance/fix/quality) branches and attach them to pre, but I've always had a problem with this in that for some reason, already-pushed commits to Main/Master tend to duplicate themselves and then I have to edit the commit history which can be a pretty big pain if you don't know what you're doing.

**Comment by Coedice** [comment]
This is the way we do it at my workplace, which I'm a fan of:
1. We protect the main branch so that the only way code goes to main is through PRs. Each PR has a assignee (the person who does the code change), and a reviewer.
2. For a PR to be merged, the reviewer must approve the PR, and then the assignee is the person who clicks the merge button when they are ready. This is like two people having to turn their keys for the change to go through.
3. Reviewers leave their review comments, if any, on the PR. [Conventional comments](https://conventionalcomments.org/) is a popular review format. Once all blocking review comments are made, they can approve the PR.
4. Each branch has only one contributor, this way you can feel free to use what would otherwise be dangerous commands such as `git push -f` or `git rebase --interactive` since you know they won't affect anyone else.

Those are probably the main principles. What do you think?

**Reply by njsch** [reply]
I have set up a [branch protection rule](https://github.com/eduvis/eduvis.github.io/settings/branches) for [main](https://github.com/eduvis/eduvis.github.io/tree/main). Does this satisfy your suggestion?

**Comment by Coedice** [comment]
Yes, I think that works 👍🏻
